main/src/ca/uwaterloo/flix/language/phase/unification/set/SetFormula.scala

```diff
@@ -609,122 +582,34 @@ object SetFormula {
     * Nested unions are put into a single union.
     */
   def mkUnionAll(fs: List[SetFormula]): SetFormula = {
-    // We need to do two things:
-    // - Separate subformulas into specific buckets of pos/neg elements/variables/constants and other.
-    // - Flatten nested unions.
+    def visit(l: List[SetFormula], elmAcc: SortedSet[Int], seenCsts: Set[Int], seenVars: Set[Int]): List[SetFormula] = l match {
+      case Nil => if (elmAcc.isEmpty) Nil else ElemSet(elmAcc) :: Nil
 
-    // es1 ∪ !es2 ∪ es3 ∪ !es4 ∪ ..
-    // empty ∪ es1 ∪ !es2 ∪ es3 ∪ !es4 ∪ ..       (neutral union formula)
-    // = (empty ∪ es1 ∪ es3 ∪ !es2 ∪ !es4) ∪ ..   (intersection associativity)
-    // = !!(empty ∪ es1 ∪ es3 ∪ !es2 ∪ !es4) ∪ .. (double complement)
-    // = !(univ ∩ !es1 ∩ !es3 ∩ es2 ∩ es4) ∪ .. (double complement)
-    // So we keep one CofiniteIntSet that represents both element sets -
-    // differencing positive elems and intersecting negative elements.
+      case ElemSet(s) :: rs => visit(rs, elmAcc ++ s, seenCsts, seenVars)
 
-    val cstsPos = mutable.Set.empty[Cst]
-    val varsPos = mutable.Set.empty[Var]
-    var elemNeg0 = CofiniteIntSet.universe
-    val cstsNeg = mutable.Set.empty[Cst]
-    val varsNeg = mutable.Set.empty[Var]
-    val other = mutable.ListBuffer.empty[SetFormula]
+      case (f@Cst(c)) :: rs =>
+        if (seenCsts.contains(c))
+          visit(rs, elmAcc, seenCsts, seenVars)
+        else
+          f :: visit(rs, elmAcc, seenCsts + c, seenVars)
```

Comment:
I think it'd be nice with another accumulator to make it stack safe

Reply:
I would wait to see if it is a problem. Most of our functions are not stack safe :[

Reply:
yeah that's why I thought It'd be nice